source/common/router/config_impl.h

```diff
@@ -547,6 +547,60 @@ class InternalRedirectPolicyImpl : public InternalRedirectPolicy {
 };
 using DefaultInternalRedirectPolicy = ConstSingleton<InternalRedirectPolicyImpl>;
 
+// Manages various optional route timeouts. We pack them in a separate data structure for memory
+// efficiency:
+// - Sharing one bool avoids the overhead of `absl::optional`, saving 8 bytes per timeout
+// - If none of the timeouts are configured, we avoid tracking any of the `std::duration` values,
+// saving another 8 bytes per timeout
+class OptionalTimeouts {
+public:
+  OptionalTimeouts(const envoy::config::route::v3::RouteAction& route);
+
+  absl::optional<std::chrono::milliseconds> idleTimeout() const {
+    return has_idle_timeout_ ? absl::optional<std::chrono::milliseconds>(idle_timeout_)
+                             : absl::nullopt;
+  }
+  absl::optional<std::chrono::milliseconds> maxStreamDuration() const {
+    return has_max_stream_duration_
+               ? absl::optional<std::chrono::milliseconds>(max_stream_duration_)
+               : absl::nullopt;
+  }
+  absl::optional<std::chrono::milliseconds> grpcTimeoutHeaderMax() const {
+    return has_grpc_timeout_header_max_
+               ? absl::optional<std::chrono::milliseconds>(grpc_timeout_header_max_)
+               : absl::nullopt;
+  }
+  absl::optional<std::chrono::milliseconds> grpcTimeoutHeaderOffset() const {
+    return has_grpc_timeout_header_offset_
+               ? absl::optional<std::chrono::milliseconds>(grpc_timeout_header_offset_)
+               : absl::nullopt;
+  }
+  absl::optional<std::chrono::milliseconds> maxGrpcTimeout() const {
+    return has_max_grpc_timeout_ ? absl::optional<std::chrono::milliseconds>(max_grpc_timeout_)
+                                 : absl::nullopt;
+  }
+  absl::optional<std::chrono::milliseconds> grpcTimeoutOffset() const {
+    return has_grpc_timeout_offset_
+               ? absl::optional<std::chrono::milliseconds>(grpc_timeout_offset_)
+               : absl::nullopt;
+  }
+
+private:
+  std::chrono::milliseconds idle_timeout_;
+  std::chrono::milliseconds max_stream_duration_;
```

Comment:
You can go further define them as unique_ptr<T>, which takes much less space than "long long".

I think you could also just define them as uint32_t, chrono::milliseconds is trivially constructible from int.

Reply:
milliseconds and unique_ptr are both 8 bytes, there is nothing to gain from the unique_ptr.

Reply:
agreed; but maybe drop a TODO because I'm not sure we really need to represent more than 4 billion milliseconds (about 46 days) for these timeouts; so you could hold them in this structure as uint32_t if you want to save some memory.

Maybe you don't want to do that in this PR since that would be a semantic change.

Reply:
I vaguely recall there is some optimization that makes unique_ptr 0 cost when it's empty. 

The other way to save would be to use array with mask but probably not worth the extra complexity.

Reply:
Right - that's what I was not suggesting doing that in this PR but it might be good to drop a TODO or just queue it onto the issue where you are tracking potential optimizations.

Reply:
@pradeepcrao is already working on a follow up improvement here. 

I don't think a TODO to switch to int32s would be very productive. Given that the current proto config supports int64 values, it would be a hard change to make. There are probably users out there that had set it to int64::max as a way of saying "no timeout".